File: ac_filter.m
function [inputAC,outlierNum] = ac_filter(inputAC) %#codegen

coder.inline('never')

% threshold for outlier filter
thresholdFactor = single(0.6);
maxBufferLength = 300;

persistent forwardBuffer;
persistent reverseBuffer;

if isempty(forwardBuffer)
    forwardBuffer = zeros(maxBufferLength, 1, 'single');
end
if isempty(reverseBuffer)
    reverseBuffer = zeros(maxBufferLength, 1, 'single');
end

% use customized filloutliers function
[inputAC, outlierNum] = fill_outliers(inputAC, thresholdFactor);

% use zero-phase digital filter
% a, b are parameters for butter bandpass filter
% [b, a] = butter(3, [0.5/(0.5*50), 4/(0.5*50)], 'bandpass');
a = single([1,-5.04456767,10.69467700,-12.21737356,7.94145633,-2.78601016,0.41183913]);
b = single([0.00716766,0,-0.02150300,0,0.02150300,0,-0.00716766]);
sampleCount = length(inputAC);
filterOrder = length(a);

for sampleIdx = 1:sampleCount
    accValue = single(0);

    for coeffIdx = 1:filterOrder
        inputIdx = sampleIdx - coeffIdx + 1;
        if inputIdx >= 1
            accValue = accValue + b(coeffIdx) * inputAC(inputIdx);
        end
    end

    for coeffIdx = 2:filterOrder
        outputIdx = sampleIdx - coeffIdx + 1;
        if outputIdx >= 1
            accValue = accValue - a(coeffIdx) * forwardBuffer(outputIdx);
        end
    end

    forwardBuffer(sampleIdx) = accValue;
end

for sampleIdx = 1:sampleCount
    reverseBuffer(sampleIdx) = forwardBuffer(sampleCount - sampleIdx + 1);
end

for sampleIdx = 1:sampleCount
    accValue = single(0);

    for coeffIdx = 1:filterOrder
        inputIdx = sampleIdx - coeffIdx + 1;
        if inputIdx >= 1
            accValue = accValue + b(coeffIdx) * reverseBuffer(inputIdx);
        end
    end

    for coeffIdx = 2:filterOrder
        outputIdx = sampleIdx - coeffIdx + 1;
        if outputIdx >= 1
            accValue = accValue - a(coeffIdx) * forwardBuffer(outputIdx);
        end
    end

    forwardBuffer(sampleIdx) = accValue;
end

for sampleIdx = 1:sampleCount
    inputAC(sampleIdx) = forwardBuffer(sampleCount - sampleIdx + 1);
end

end

File: codegen/sleepon0422/ac_filter.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * ac_filter.c
 *
 * Code generation for function 'ac_filter'
 *
 */

/* Include files */
#include "ac_filter.h"
#include "fill_outliers.h"
#include "filter.h"
#include "ppg_process_data.h"
#include "rt_nonfinite.h"
#include <string.h>

/* Function Definitions */
double ac_filter(float inputAC[150])
{
  double outlierNum;
  float b_fv[150];
  float b_inputAC[150];
  int i;
  /*  threshold for outlier filter */
  /*  use customized filloutliers function */
  outlierNum = fill_outliers(inputAC);
  /*  use zero-phase digital filter */
  /*  a, b are parameters for butter bandpass filter */
  /*  [b, a] = butter(3, [0.5/(0.5*50), 4/(0.5*50)], 'bandpass'); */
  /*  Forward filter */
  /*  Reverse and filter */
  /*  reverse and output */
  filter(fv, fv1, inputAC, b_inputAC);
  for (i = 0; i < 150; i++) {
    b_fv[i] = b_inputAC[149 - i];
  }
  filter(fv, fv1, b_fv, inputAC);
  for (i = 0; i < 150; i++) {
    b_inputAC[i] = inputAC[149 - i];
  }
  memcpy(&inputAC[0], &b_inputAC[0], 150U * sizeof(float));
  return outlierNum;
}

double b_ac_filter(float inputAC_data[], int *inputAC_size)
{
  double outlierNum;
  float b_forward_filtered_data[300];
  float forward_filtered_data[300];
  int forward_filtered;
  int forward_filtered_size;
  int i;
  /*  threshold for outlier filter */
  /*  use customized filloutliers function */
  outlierNum = b_fill_outliers(inputAC_data, inputAC_size);
  /*  use zero-phase digital filter */
  /*  a, b are parameters for butter bandpass filter */
  /*  [b, a] = butter(3, [0.5/(0.5*50), 4/(0.5*50)], 'bandpass'); */
  /*  Forward filter */
  forward_filtered_size =
      b_filter(inputAC_data, *inputAC_size, forward_filtered_data);
  /*  Reverse and filter */
  forward_filtered = forward_filtered_size - 1;
  for (i = 0; i <= forward_filtered; i++) {
    b_forward_filtered_data[i] =
        forward_filtered_data[(forward_filtered_size - i) - 1];
  }
  if (forward_filtered_size - 1 >= 0) {
    memcpy(&forward_filtered_data[0], &b_forward_filtered_data[0],
           (unsigned int)forward_filtered_size * sizeof(float));
  }
  *inputAC_size =
      b_filter(forward_filtered_data, forward_filtered_size, inputAC_data);
  /*  reverse and output */
  forward_filtered_size = *inputAC_size - 1;
  forward_filtered = forward_filtered_size + 1;
  for (i = 0; i <= forward_filtered_size; i++) {
    forward_filtered_data[i] = inputAC_data[forward_filtered_size - i];
  }
  *inputAC_size = forward_filtered_size + 1;
  if (forward_filtered - 1 >= 0) {
    memcpy(&inputAC_data[0], &forward_filtered_data[0],
           (unsigned int)forward_filtered * sizeof(float));
  }
  return outlierNum;
}

/* End of code generation (ac_filter.c) */
